crates/polars-plan/src/logical_plan/optimizer/predicate_pushdown/utils.rs

```diff
@@ -409,29 +331,69 @@ where
     local_predicates
 }
 
-/// predicates that need the full context should not be pushed down to the scans
-/// example: min(..) == null_count
-pub(super) fn partition_by_full_context(
-    acc_predicates: &mut PlHashMap<Arc<str>, Node>,
-    expr_arena: &Arena<AExpr>,
-) -> Vec<Node> {
-    // TODO!
-    // Assert that acc_predicates does not contain a mix of groups sensitive and
-    // non-groups sensitive predicates, as this should have been handled
-    // earlier under push_down::match::Selection.
-    if acc_predicates.values().any(|node| {
-        has_aexpr(*node, expr_arena, |ae| match ae {
-            AExpr::BinaryExpr { left, right, .. } => {
-                expr_arena.get(*left).groups_sensitive()
-                    || expr_arena.get(*right).groups_sensitive()
-            },
+/// An expression blocks predicates from being pushed past it if its results for
+/// the subset where the predicate evaluates as true becomes different compared
+/// to if it was performed before the predicate was applied. This is in general
+/// any expression that produces outputs based on groups of values
+/// (i.e. groups-wise) rather than individual values (i.e. element-wise).
+///
+/// Examples of expressions whose results would change, and thus block push-down:
+/// - any aggregation - sum, mean, first, last, min, max etc.
+/// - sorting - as the sort keys would change between filters
+pub(super) fn aexpr_blocks_predicate_pushdown(node: Node, expr_arena: &Arena<AExpr>) -> bool {
+    let mut stack = Vec::<Node>::with_capacity(4);
+    stack.push(node);
+
+    // Cannot use `has_aexpr` because we need to ignore any literals in the RHS
+    // of an `is_in` operation.
+    while let Some(node) = stack.pop() {
+        let ae = expr_arena.get(node);
+
+        if match ae {
+            // These literals do not come from the RHS of an is_in, meaning that
+            // they are projected as either columns or predicates, both of which
+            // rely on the height of the dataframe at this level and thus need
+            // to block pushdown.
+            AExpr::Literal(LiteralValue::Range { .. }) => true,
+            AExpr::Literal(LiteralValue::Series(s)) => s.len() > 1,
             ae => ae.groups_sensitive(),
-        })
-    }) {
-        let local_predicates = acc_predicates.values().copied().collect::<Vec<_>>();
-        acc_predicates.clear();
-        local_predicates
-    } else {
-        vec![]
+        } {
+            return true;
+        }
+
+        match ae {
+            #[cfg(feature = "is_in")]
+            AExpr::Function {
+                function: FunctionExpr::Boolean(BooleanFunction::IsIn),
+                input,
+                ..
+            } => {
+                // Handles a special case where the expr contains a series, but it is being
+                // used as part of an `is_in`, so it can be pushed down.
+                let values = input.get(1).unwrap();
```

Comment:
Can we branch here instead of `unwrap`. There are cases where we don't have input expressions.